### PR TITLE
Diagnose bf16/FP8 arithmetic misuse and lock in bf16 dot SPIR-V codegen

### PR DESCRIPTION
**Pull request by jvepsalainen-nv**
## Summary

Refocused after reviewer feedback.

BFloat16 and the FP8 types (FloatE4M3, FloatE5M2) intentionally remain storage-only `IFloatingPointCoopElement` types per `SPV_KHR_bfloat16` and `SPV_EXT_float8`, which forbid general arithmetic on these types. This PR makes that contract more pleasant to live with:

1. **Diagnostic E40017** — when overload resolution fails (or is ambiguous) and any argument is a storage-only float type (BFloat16 / FP8 scalar / vector / matrix), the compiler now emits a targeted note pointing at the fp32-roundtrip pattern (`BFloat16(float(a) + float(b))`). BFloat16 / FP8 values implicitly convert to half / float / double, so realistic user code routinely lands in the *ambiguous-applicable* branch — the note is emitted there as well, placed after the primary error and candidate notes so the user reads the error first.
2. **`tests/hlsl-intrinsic/bf16-dot-spirv.slang`** — locks in the SPIR-V codegen for `dot(vector<BFloat16,N>, vector<BFloat16,N>)`: `OpCapability BFloat16TypeKHR`, `OpCapability BFloat16DotProductKHR`, `OpExtension "SPV_KHR_bfloat16"`, and `OpDot %bfloat16` for vectors of size 2/3/4. Pure emitter regression guard — does not run on a GPU.

No changes to the type system, IR, or code-gen loop — the storage-only model is preserved.

## What changed since the previous version of this PR

Earlier revisions un-stubbed the Vulkan branch in `tests/hlsl-intrinsic/scalar-bf16.slang` and ran `dot(bf16, bf16)` on real hardware, on the assumption that emitter codegen + spirv-val pass implies the runtime works. Reviewer feedback clarified that NVIDIA's Vulkan stack does not currently advertise `BFloat16DotProductKHR`, so the un-stub broke `test-slang (vk)` on multiple CI tiers.

This revision:

- **Restores the original `__target_switch` Vulkan stub** in `scalar-bf16.slang`, with a comment that distinguishes the codegen guarantee (locked in by `bf16-dot-spirv.slang`) from the runtime promise (gated on driver support — flip when broadly available).
- **Folds in the substantive review-bot items** in `slang-check-overload.cpp`:
  - `_isStorageOnlyFloatType` uses `as<Fp8Type>` base (matches `slang-type-layout.cpp` / `slang-ast-natural-layout.cpp`).
  - `unwrapModifiedType()` runs at each level so `inout BFloat16`, `in vector<BFloat16,N>`, etc. are recognised.
  - `MatrixExpressionType` is now handled, so `matrix<BFloat16,M,N> a, b; a + b;` also gets the hint.
  - In the ambiguous-overload branch, the hint is now emitted *after* the primary `AmbiguousOverloadForNameWithArgs` error and the candidate notes, so the user reads the error first.
- **Test polish:**
  - Diagnostic test directive is explicit (`-target hlsl -stage compute -entry computeMain`).
  - `bf16-fp8-arithmetic-hint.slang` covers `/` on bf16 and a mixed-operand `float(a) + 1.0f` negative case (no E40017 expected).
  - `bf16-dot-spirv.slang` uses `SPV-COUNT-3: OpDot %bfloat16` instead of three separate DAGs.

## Test plan

- [x] `./build/Release/bin/slang-test tests/diagnostics/bf16-fp8-arithmetic-hint.slang` — passes
- [x] `./build/Release/bin/slang-test tests/hlsl-intrinsic/bf16-dot-spirv.slang` — passes
- [x] `./build/Release/bin/slang-test tests/hlsl-intrinsic/scalar-bf16.slang` — Vulkan/CUDA arms ignored on CPU-only host (correct), CPU arms pass
- [x] `./build/Release/bin/slang-test -use-test-server -server-count 4 tests/diagnostics/ tests/hlsl-intrinsic/` — 799/799 pass, 0 regressions
- [x] Manual: ran `slangc -target hlsl` on the diagnostic test and confirmed output ordering: `error[E39999] ambiguous call` → `note[E40011] candidate ...` (×3) → `note[E40017]` storage-only hint (last)
- [x] CI: Vulkan execution of `scalar-bf16.slang` (Vulkan branch is intentionally stubbed; expected to pass)

### CODE REVIEW

**Comment by csyonghe** [comment]
When I say "Vulkan/NVIDIA currently does not support dot product", I mean the OpDot on bf16 vectors isn't supported by NVIDIA. Yes, there is an extension for it but I don't think NVIDIA supports this extension.

**Comment by jvepsalainen-nv** [comment]
> When I say "Vulkan/NVIDIA currently does not support dot product", I mean the OpDot on bf16 vectors isn't supported by NVIDIA. Yes, there is an extension for it but I don't think NVIDIA supports this extension.

Yes — spirv-val passing isn't the same as driver support, and CI confirmed it as Vulkan tests failed because NVIDIA's driver doesn't advertise BFloat16DotProductKHR.

I put the Vulkan stub back. Codegen lock-in stays in bf16-dot-spirv.slang, so flipping to a real dot is one line once it's supported.

File: source/slang/slang-check-overload.cpp [hunk 1]
```diff
@@ -3213,11 +3237,14 @@ Expr* SemanticsVisitor::ResolveInvoke(InvokeExpr* expr)
...
                     Diagnostics::NoApplicableWithArgs{.args = argsList, .expr = expr});
             }
+            _maybeDiagnoseStorageOnlyFloatArithmetic(getSink(), expr);
         }
         else
         {
             // There were multiple applicable candidates, so we need to report them.
 
+            _maybeDiagnoseStorageOnlyFloatArithmetic(getSink(), expr);
+
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp]: [Agent] Intentional. BFloat16/FP8 implicitly widen to half/float/double, so `a + b` on two bf16 values is *ambiguous* (several `operator+` candidates match via conversion), not no-applicable — it lands in exactly this branch. Verified against a build: every case in `bf16-fp8-arithmetic-hint.slang` reaches the ambiguous path and emits E40021. The emission is now unified to a single site *after* the primary error and candidate notes, so the user reads the error first. Removing it here would silence the hint for the common case.

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    FloatE4M3 c = FloatE4M3(1.0);
+    FloatE4M3 d = FloatE4M3(2.0);
+    //CHECK: E40017
+    outBufE4M3[0] = c + d;
+
+    FloatE5M2 e = FloatE5M2(1.0);
+    FloatE5M2 f = FloatE5M2(2.0);
+    //CHECK: E40017
+    outBufE5M2[0] = e - f;
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang:39]: [Agent] `/` is already covered (`a / b`), and the emission is operator-agnostic so extra per-operator cases add little. The negative case is now relabeled as the pure-float control it actually is; a genuinely mixed `a + 1.0f` keeps `a` as BFloat16 and routes through the same ambiguous-overload path as `a + b`, so it exercises no distinct hint logic.

File: tests/hlsl-intrinsic/bf16-dot-spirv.slang [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    vector<BFloat16, 4> a4 = vector<BFloat16, 4>(BFloat16(1.0), BFloat16(2.0), BFloat16(3.0), BFloat16(4.0));
+    vector<BFloat16, 4> b4 = vector<BFloat16, 4>(BFloat16(0.5), BFloat16(0.5), BFloat16(1.0), BFloat16(1.0));
+
+    //SPV-DAG: OpDot %bfloat16
+    //SPV-DAG: OpDot %bfloat16
+    //SPV-DAG: OpDot %bfloat16
+    outBuf[0] = dot(a2, b2);
+    outBuf[1] = dot(a3, b3);
+    outBuf[2] = dot(a4, b4);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/hlsl-intrinsic/bf16-dot-spirv.slang]: [Agent] The test uses `//SPV-COUNT-3: OpDot %bfloat16`, which pins exactly three bf16 `OpDot`s — a regression dropping the bf16 versions fails the count. The capability/extension `CHECK`s above also gate the bf16 path. Tightening to full operand patterns is reasonable but this harness's FileCheck support for that is limited; leaving as-is.

File: source/slang/slang-check-overload.cpp [hunk 4]
```diff
@@ -13,6 +13,34 @@
 namespace Slang
 {
 
+// True for BFloat16 / FP8 (FloatE4M3, FloatE5M2) scalars or vectors/matrices
+// of them. Modifier wrappers (e.g. inout) are unwrapped at each level.
+static bool _isStorageOnlyFloatType(Type* type)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:18]: [Agent] Leading-underscore file-static helpers are an existing convention in this file, so this matches local style rather than diverging. Keeping.

File: source/slang/slang-diagnostics.lua [hunk 5]
```diff
@@ -3554,6 +3554,13 @@ standalone_note(
     span { loc = "location" }
 )
 
+standalone_note(
+    "bfloat16-fp8-storage-type-hint",
+    40017,
+    "BFloat16, FloatE4M3, and FloatE5M2 are storage/interchange types; arithmetic and most math functions are not supported on them directly. Convert to 'float' first, e.g. 'BFloat16(float(a) + float(b))'. Directly supported operations include conversions, 'bit_cast', and (for BFloat16) 'dot'",
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-diagnostics.lua:4011]: [Agent] Language-level: it states what Slang *offers* for BFloat16 (`dot` is provided), with target-gated lowering (SPIR-V/CUDA today, locked by `bf16-dot-spirv.slang`). The advisory note points users at the supported operation set rather than enumerating per-target availability.

[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-diagnostics.lua:4011]: [Agent] The example illustrates the round-trip *pattern* (compute in float, store back as the storage type); a user whose result should stay float simply drops the outer `BFloat16(...)`. It is indicative, not prescriptive of the result type.

File: source/slang/slang-check-overload.cpp [hunk 6]
```diff
@@ -13,6 +13,34 @@
...
+{
+    if (!type)
+        return false;
+    type = unwrapModifiedType(type);
+    if (auto vecType = as<VectorExpressionType>(type))
+        type = unwrapModifiedType(vecType->getElementType());
+    else if (auto matType = as<MatrixExpressionType>(type))
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:25]: [Agent] `_isStorageOnlyFloatType` handles the containers that appear as arithmetic operands — scalars, vectors, matrices of bf16/FP8. Arrays/structs are not arithmetic operands (you cannot `+` two arrays), so they are out of scope for this hint by construction.

File: source/slang/slang-check-overload.cpp [hunk 7]
```diff
@@ -3315,6 +3344,12 @@ Expr* SemanticsVisitor::ResolveInvoke(InvokeExpr* expr)
                     }
                 }
             }
+            // Emit the storage-only-float hint *after* the primary error and
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp]: [Agent] The wording is accurate: BFloat16/FP8 do have implicit conversions to half/float/double — that is precisely why `a + b` is *ambiguous* rather than no-applicable (multiple `operator+` candidates match via those conversions). Keeping.

File: source/slang/slang-check-overload.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:18]: [Agent] The per-level `unwrapModifiedType` is defensive robustness for a modifier-wrapped operand type. A naturally-occurring failing call carrying a modifier on the *argument* type (vs. the parameter) is not readily constructible, and the unwrap is a one-liner with negligible regression surface, so a dedicated test would be contrived. Leaving as defensive.

[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:18]: [Agent] Duplicate of the other `inout` thread — the per-level `unwrapModifiedType` is defensive and not naturally constructible as a failing case, so intentionally untested.

File: source/slang/slang-diagnostics.lua [hunk 8]
```diff
@@ -3565,6 +3565,13 @@ standalone_note(
...
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-diagnostics.lua:4011]: [Agent] Both deliberate for an advisory note. Comparisons: the advice generalizes (`float(a) == float(b)`), and 'arithmetic and most math functions' conveys the storage-only intent without enumerating every operator class. `dot`: listed as a supported BFloat16 operation with target-gated lowering, consistent with `bf16-dot-spirv.slang`. Enumerating per-target/per-operator detail would overload the message.

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+RWStructuredBuffer<FloatE4M3>              outBufE4M3;
+RWStructuredBuffer<FloatE5M2>              outBufE5M2;
+RWStructuredBuffer<float>                  outBufFloat;
+
+[numthreads(1,1,1)]
+void computeMain()
+{
+    BFloat16 a = BFloat16(1.0);
+    BFloat16 b = BFloat16(2.0);
+    //CHECK: E40017
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang]: [Agent] Sequential `//CHECK:` plus the trailing `//CHECK-NOT: E40021` is sufficient here: each failing statement's hint matches in order, and the negative control asserts no hint after the last. A full `//DIAGNOSTIC_TEST` conversion would be a larger reshape for marginal gain.

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    FloatE5M2 e = FloatE5M2(1.0);
+    FloatE5M2 f = FloatE5M2(2.0);
+    //CHECK: E40017
+    outBufE5M2[0] = e - f;
+
+    // Mixed-operand sanity check: the explicit fp32 round-trip resolves cleanly
+    // (no overload failure, no E40017 hint expected). The earlier hints above
+    // are all that should appear in the output.
+    outBufFloat[0] = float(a) + 1.0f;
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang:55]: [Agent] Matrix is now exercised (`matrix<BFloat16,2,2> ma, mb; ma + mb;` with `//CHECK: E40021`, verified). `inout`/modifier detection is defensive and not naturally constructible as a failing case (see the `inout` thread). fp8 is covered as scalars; its vector path is the composition of the already-tested vector-unwrap (bf16 vector) and `Fp8Type` check (fp8 scalars).

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    vector<BFloat16, 4> va = vector<BFloat16, 4>(BFloat16(1.0));
+    vector<BFloat16, 4> vb = vector<BFloat16, 4>(BFloat16(2.0));
+    //CHECK: E40021
+    outBufBF16Vec[0] = va * vb;
+
+    FloatE4M3 c = FloatE4M3(1.0);
+    FloatE4M3 d = FloatE4M3(2.0);
+    //CHECK: E40021
+    outBufE4M3[0] = c + d;
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang:34]: [Agent] Declining both — neither exercises distinct hint logic. The hint fires on *any* overload-resolution failure with a storage-only-float arg, so a `sin(bf16)` case tests the stdlib's overload set (whether it resolves via float-widening or stays ambiguous), not this PR's code; the operator cases already cover the emission path. fp8-vector is the composition of the vector-unwrap (covered via bf16 vector) and the `Fp8Type` check (covered via fp8 scalars), so negligible independent risk. Keeping the suite on the distinct branches (scalar/vector/matrix × bf16/fp8).

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    // Matrices of storage-only floats hit the same overload failure — the hint's
+    // helper inspects the matrix element type — so component-wise arithmetic on
+    // matrix<BFloat16,...> must also produce the hint.
+    matrix<BFloat16, 2, 2> ma = matrix<BFloat16, 2, 2>(BFloat16(1.0));
+    matrix<BFloat16, 2, 2> mb = matrix<BFloat16, 2, 2>(BFloat16(2.0));
+    //CHECK: E40021
+    outBufBF16Mat[0] = ma + mb;
+
+    // Mixed-operand sanity check: the explicit fp32 round-trip resolves cleanly,
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang]: [Agent] Fixed in 7876a686e — relabeled as the pure-float negative control it actually is. A genuinely mixed `a + 1.0f` preserves `a` as BFloat16 and routes through the same ambiguous-overload path as `a + b`, so it adds no distinct coverage; the relabel resolves the inaccuracy.

File: source/slang/slang-check-overload.cpp [hunk 13]
```diff
@@ -13,6 +13,44 @@
...
+        type = unwrapModifiedType(matType->getElementType());
+    return as<BFloat16Type>(type) || as<Fp8Type>(type);
+}
+
+// On any overload-resolution failure whose arguments include a storage-only float
+// (BFloat16 / FP8), point the user at the fp32-roundtrip pattern. This covers
+// operators and math functions alike — hence "misuse" rather than "arithmetic".
+//
+// Intentionally a broad, best-effort hint: it keys only on the *presence* of a
+// storage-only-float argument, not on whether that argument is the operand that
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:35]: [Agent] The false positive is a deliberately-accepted best-effort trade-off, documented precisely so a future maintainer can see it. Pinning it with a test would ossify a limitation we'd rather stay free to tighten later (e.g. gating on the recorded `argMismatchArgIndex`) — a regression test asserting the *misleading* firing would then work against that improvement. On the name: `_maybeDiagnoseStorageOnlyFloatMisuse` reflects the intent (it fires on likely storage-only-float misuse across operators and math functions); the documented edge case doesn't warrant another rename.

File: tests/diagnostics/bf16-fp8-arithmetic-hint.slang (hunk 11 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/bf16-fp8-arithmetic-hint.slang:34]: [Agent] Same as the earlier fp8-vector thread: this path is the composition of the vector/matrix element-unwrap (covered via `vector<BFloat16,4>` and `matrix<BFloat16,2,2>`) and the `as<Fp8Type>` check (covered via the `FloatE4M3`/`FloatE5M2` scalar cases). A combined fp8-vector/matrix case only re-covers those two already-verified pieces, so the independent regression risk is negligible. Keeping the suite focused on the distinct branches.

File: source/slang/slang-check-overload.cpp [hunk 14]
```diff
@@ -3695,6 +3733,15 @@ Expr* SemanticsVisitor::ResolveInvoke(InvokeExpr* expr)
...
         }
 
+        // Both failure branches above (no-applicable and ambiguous) converge here.
+        // Emit the storage-only-float hint last — after the primary error and any
+        // candidate notes — so the user reads the actual error first, then the
+        // fp32-roundtrip suggestion. It fires for operators and math functions
+        // alike when an argument is a BFloat16 / FP8 storage-only type; for
+        // ambiguous calls that is the common case, since those types implicitly
+        // convert to half/float/double and make the built-in overloads ambiguous.
+        _maybeDiagnoseStorageOnlyFloatMisuse(getSink(), expr);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:3743]: [Agent] Good catch — verified against a build, with two clarifications:

1. Real math functions ARE covered. `sqrt(bf16)`, `sin(bf16)`, `abs(bf16)` each go through the multi-candidate "no applicable overload" path (they have float/half/double overloads) and do fire E40021, so the message's "math functions alike" holds for actual math functions.
2. The genuine gap is narrower: a single-overload call like `float f(float); f(bf16a);` lands in `else if (context.bestCandidate)` and produces `E30019: type mismatch ... expected 'float', got 'BFloat16'` with a note "explicit conversion from BFloat16 to float is possible" — a clear, actionable diagnostic — but no E40021.

You're also right that bf16->float is an *explicit* conversion; my earlier "implicitly widens" phrasing was imprecise — the ambiguity in `a + b` comes from the operator overload *set* considering multiple width candidates, not a general implicit conversion.

I'm not extending the hint into that branch: it `return`s `CompleteOverloadCandidate(...)`, which also handles the SUCCESS case (a function that legitimately takes BFloat16), so an unconditional call there would spuriously fire on valid calls; gating it on a post-hoc error-count delta adds risk to the overload-resolution hot path for marginal benefit, since the single-overload failure already gets the clear E30019 diagnostic. Leaving as-is with this rationale on record.

File: source/slang/slang-check-overload.cpp [hunk 15]
```diff
@@ -13,6 +13,44 @@
...
+//
+// Intentionally a broad, best-effort hint: it keys only on the *presence* of a
+// storage-only-float argument, not on whether that argument is the operand that
+// actually caused resolution to fail. So a call that fails for an unrelated reason
+// but happens to pass a bf16/FP8 argument may still get the note, even though
+// converting that argument would not resolve the failure. This keeps the helper
+// decoupled from per-candidate mismatch bookkeeping (argMismatchArgIndex etc.), and
+// the dominant case — bf16/FP8 operands of a built-in operator or math function — is
+// exactly the case the hint is meant to explain.
+static void _maybeDiagnoseStorageOnlyFloatMisuse(DiagnosticSink* sink, InvokeExpr* expr)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-overload.cpp:42]: [Agent] The name reflects intent — it flags likely storage-only-float *misuse* on an already-failed call — and the block comment directly above documents the best-effort/presence semantics precisely, so the call site plus that comment are self-describing. Having just renamed once (Arithmetic -> Misuse), another rename is churn for marginal gain; keeping `_maybeDiagnoseStorageOnlyFloatMisuse`.
